add validation rule for visualize

graceguo-supercat opened the pull request:
In Visualize modal, even user didn't select any columns and clicks Visualize button, we still send request to server-side, and display a warning message returned from server-side:
![error message from server-side](http://g.recordit.co/sqQ03JzXhz.gif)

Instead of server-side validation, we can add a client-side validation rule to make sure at least one matrix is selected (based on the user selected chart type).

Commit:
redux visualize modal

Commit:
redux visualize modal

- apply redux
- add unit test for connect container component
- fix lint error

Commit:
redux visualize modal

- apply redux
- add unit test for connect container component
- fix lint error

Commit:
add unit tests for VisuaizeModal

Commit:
add validation rule

Commit:
add validation rule

Commit:
add unit tests for VisualizeModal component

vavrusa committed:
filter_box: allow creatable entries (#2804)

The reason is that it should be able to create arbitrary filters
over attributes with very high cardinality, where loading all
possible combinations into the filter box is infeasible,
e.g. IPv6 addresses.

Commit:
Merge remote-tracking branch 'supercat/gg-addTestCoverage-VisualizeModal' into gg-addTestCoverage-VisualizeModal

# Conflicts:
#	superset/assets/javascripts/SqlLab/actions.js

Commit:
[clarity/consistency] rename /explorev2/ -> /explore/ (#2802)

* rename /explorev2/ -> /explore/

* add redirect for existing explorev2 urls

* fix long line

* remove extra line

* fix missed ref in spec

Commit:
remove changes on css file (merge by mistake)

Commit:
remove changes on css file (merge by mistake)

Commit:
update unit test description

setStateFromProps required valid props as parameter, otherwise it should not change state.

Commit:
Merge remote-tracking branch 'upstream/master' into gg-addValidationRuleForVisualize

ascott commented:
is this the before or after gif?

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -12,10 +12,10 @@ import { getExploreUrl } from '../../explore/exploreUtils';
-  { value: 'pie', label: 'Pie Chart', requiresTime: false },
-  { value: 'line', label: 'Time Series - Line Chart', requiresTime: true },
-  { value: 'bar', label: 'Time Series - Bar Chart', requiresTime: true },
+  { value: 'dist_bar', label: 'Distribution - Bar Chart', requiresTime: false, requiresMatrix: true },
```
rather than calling this `requiresMatrix`, would it be more clear to specify `requiresDimension` and `requiresAggregrateFn`?

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -12,10 +12,10 @@ import { getExploreUrl } from '../../explore/exploreUtils';
 import * as actions from '../actions';
 
 const CHART_TYPES = [
```
i think we should move this to the `visTypes` constant in the store so we are not duplicating the labels, and can make use of these additional flags elsewhere as well.
https://github.com/airbnb/superset/blob/master/superset/assets/javascripts/explore/stores/visTypes.js#L75

mistercrunch committed:
[explore] include ControlHeader as part of Control interface (#2809)

* [explore] include ControlHeader as part of Control interface

* Adressing comments

mistercrunch committed:
[explore] include ControlHeader as part of Control interface (#2809)

* [explore] include ControlHeader as part of Control interface

* Adressing comments

Commit:
reduce clientside timeout limit (#2820)

it was set to 60 seconds. But it seems gateway will return 504 around 50 seconds.
Will reduce client side timeout limit to 45 seconds.

Commit:
reduce clientside timeout limit (#2820)

it was set to 60 seconds. But it seems gateway will return 504 around 50 seconds.
Will reduce client side timeout limit to 45 seconds.

Commit:
add validation rule for VisualizeModal

- add check for requiresDimension
- add check for requiresAggregationFn
- use chart types defined from visTypes.js

Commit:
add validation rule for VisualizeModal

- add check for requiresDimension
- add check for requiresAggregationFn
- use chart types defined from visTypes.js

Commit:
add validation rule for VisualizeModal

- add check for requiresDimension
- add check for requiresAggregationFn
- use chart types defined from visTypes.js

Commit:
add validation rule for VisualizeModal

- add check for requiresDimension
- add check for requiresAggregationFn
- use chart types defined from visTypes.js

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -88,24 +86,34 @@ class VisualizeModal extends React.PureComponent {
-          'To use this chart type you need at least one column ' +
-          'flagged as a date');
+      if (this.state.chartType.requiresDimension) {
+        const hasMatrix = Object.keys(cols).filter(key => (cols[key].is_dim)).length > 0;
```
`hasDimension` might be more clear here rather than `hasMatrix`

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -88,24 +86,34 @@ class VisualizeModal extends React.PureComponent {
-        hints.push(
-          'To use this chart type you need at least one matrix');
+      if (this.state.chartType.requiresAggregationFn) {
+        const hasMatrix = Object.keys(cols).filter(key => (cols[key].agg)).length > 0;
```
s/hasMatrix/hasDimension

Commit:
Merge remote-tracking branch 'upstream/master' into gg-addValidationRuleForVisualize

Commit:
Merge remote-tracking branch 'supercat/gg-addValidationRuleForVisualize' into gg-addValidationRuleForVisualize

Commit:
Merge branch 'master' into gg-addValidationRuleForVisualize

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -195,6 +189,28 @@ describe('VisualizeModal', () => {
+          type: 'STRING',
+        },
+      });
+      wrapper.setState({ chartType: mockChartTypeBarChart });
```
could just use `visTypes.bar` rather than setting a var for it on L50

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -45,16 +47,8 @@ describe('VisualizeModal', () => {
-    value: 'bar',
-  };
+  const mockChartTypeBarChart = visTypes.bar;
+  const mockChartTypeTB = visTypes.line;
```
doesn't look like we use this var anywhere.

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -91,16 +89,30 @@ class VisualizeModal extends React.PureComponent {
-        hints.push(VISUALIZE_VALIDATION_ERRORS.REQUIRE_TIME);
+      if (this.state.chartType.requiresAggregationFn) {
+        const hasMatrix = Object.keys(cols).filter(key => (cols[key].agg)).length > 0;
+        if (!hasMatrix) {
```
could simplify these 2 lines to 1:
`if (Object.keys(cols).filter(key => (cols[key].agg)).length > 0) {`

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -91,16 +89,30 @@ class VisualizeModal extends React.PureComponent {
+      }
+      if (this.state.chartType.requiresDimension) {
+        const hasMatrix = Object.keys(cols).filter(key => (cols[key].is_dim)).length > 0;
+        if (!hasMatrix) {
```
could simplify to:
`if (!Object.keys(cols).filter(key => (cols[key].is_dim)).length > 0) {`

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -91,16 +89,30 @@ class VisualizeModal extends React.PureComponent {
+        let hasTime = false;
+        for (const colName in cols) {
+          const col = cols[colName];
+          if (col.hasOwnProperty('is_date') && col.is_date) {
```
why do we need both of these conditions? could we just check
`if (col.is_date) {}`

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -91,16 +89,30 @@ class VisualizeModal extends React.PureComponent {
-          hasTime = true;
+    } else {
+      if (this.state.chartType.requiresTime) {
+        let hasTime = false;
```
i think we could simplify this block by using [Array#some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some?v=example)

```js
const hasTime = cols.some(col => col.is_date);
if (!hasTime) {
  hints.push(VISUALIZE_VALIDATION_ERRORS.REQUIRE_TIME);
}
```

xiaohanyu committed:
Fix a type error in rst' sub list. (#2881)

Sub list in rst requires a blank line above and below.

Ref:
- http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#bullet-lists

justinr1234 committed:
[bugfix] allow database macro to work when CSRF is diabled (#2884)

Commit:
[sql lab] Add client-side validation before send to server

- check is_date for chart that requires time series
- check is_dim for chart that requires dimension
- check agg is defined for char that requires Aggregation function

ascott commented on superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -89,19 +88,21 @@ class VisualizeModal extends React.PureComponent {
       }
     });
+
+    const getNumColsHaveProperty = propName =>
```
would `hasColumnType` be more clear here? i would also suggest moving this to it's own method on the component.

graceguo-supercat replied:
true. So i add a helper function `hasColumnType`.

ascott commented on docs/sqllab.rst:
```diff
@@ -13,9 +13,11 @@ Feature Overview
   visualization capabilities
 - Browse database metadata: tables, columns, indexes, partitions
 - Support for long-running queries
+
```
looks like this is from another PR? maybe need to rebase master?

Commit:
add client-side validation rule

- add check for requiresDimension
- add check for requiresAggregationFn
- use chart types defined from visTypes.js

Commit:
Merge remote-tracking branch 'supercat/gg-addValidationRuleForVisualize' into gg-addValidationRuleForVisualize

# Conflicts:
#	superset/assets/javascripts/SqlLab/components/VisualizeModal.jsx

ascott commented:
this looks great! thanks for the revisions! 👍  🚢

Commit:
Merge branch 'master' into gg-addValidationRuleForVisualize

Commit:
Merge branch 'master' into gg-addValidationRuleForVisualize

Commit:
Merge branch 'master' into gg-addValidationRuleForVisualize

mistercrunch commented:
Notice: this issue has been closed because it has been inactive for 324 days. Feel free to comment and request for this issue to be reopened.
